### PR TITLE
BV Calculations Prep

### PR DESCRIPTION
**Pull request by SJuliez**
In preparation for looking at the actual BV calculations (bugs) I wanted to move out the calculation code to separate classes as for some of the entity subclasses it makes up as much as a third of the code. This PR does that.
 
- Since github just shows the BV calculation code as deleted and new, I made only the minimum changes necessary to the calculations in xyzBVCalculator (mostly replacing "this." calls with "mek." calls and using a few accessors instead of fields) to just get them compiled again after moving. So, no beauty or functional changes.
- The inheritance structure of the calculateBV() calls in Entity is cleaned up a bit. The two main methods calculateBV() and calculateBV(boolean, boolean) are now final in Entity and forward to a private method that returns the manual BV if there is one and forwards to the (abstract) actual BV calculation method if not. The actual BV calculation method isnt visible from outside and is overridden in Mech, Aero etc. where it is either very short (TeleMissile) or forwards to xyzBVCalculator.
- To the outside, calculateBV() and calculateBV(boolean, boolean) act as before.
- The "singleTrooper" BV calcuation of BA falls outside of this and is only available in BA. It's used in MML at least.

### CODE REVIEW

File: megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+import megamek.common.*;
+
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ProtoMekBVCalculator extends BVCalculator {
+
+    public static int calculateBV(Protomech protoMek, boolean ignoreSkill, StringBuffer bvText) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java:30]:
```suggestion
public class ProtoMekBVCalculator extends BVCalculator {
    public static int calculateBV(Protomech protoMek, boolean ignoreSkill, StringBuffer bvText) {
```

File: megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+
+            // don't count destroyed equipment
+            if (mounted.isDestroyed()) {
+                continue;
+            }
+
+            if (((etype instanceof WeaponType) && etype
+                    .hasFlag(WeaponType.F_AMS))
+                    || ((etype instanceof MiscType) && (etype
+                    .hasFlag(MiscType.F_ECM)
```
Comment (Windchild292) [inline megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java:91]: Same here

File: megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+        double tmmRan = Compute.getTargetMovementModifier(protoMek.getRunMP(false, true, true), false, false, protoMek.getGame()).getValue();
+        // Gliders get +1 for being airborne.
+        if (protoMek.isGlider()) {
+            tmmRan++;
+        }
+
+        final int jumpMP = protoMek.getJumpMP(false);
+        final int tmmJumped = (jumpMP > 0) ? Compute.
+                getTargetMovementModifier(jumpMP, true, false, protoMek.getGame()).getValue()
+                : 0;
```
Comment (Windchild292) [inline megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java:149]:
```
val = check
        ? a
        : b;
```
is far easier to read than this

File: megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+
+        final int umuMP = protoMek.getActiveUMUCount();
+        final int tmmUMU = (umuMP > 0) ? Compute.
+                getTargetMovementModifier(umuMP, false, false, protoMek.getGame()).getValue()
+                : 0;
```
Comment (Windchild292) [inline megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java:154]:
```
val = check
        ? a
        : b;
```
is far easier to read than this

File: megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+            bvText.append(endColumn);
+            bvText.append(startColumn);
+            bvText.append(endRow);
+
+            // add up BV of ammo-using weapons for each type of weapon,
+            // to compare with ammo BV later for excessive ammo BV rule
+            if (!(wtype.hasFlag(WeaponType.F_ENERGY)
+                    || wtype.hasFlag(WeaponType.F_ONESHOT)
+                    || wtype.hasFlag(WeaponType.F_INFANTRY) || (wtype
+                    .getAmmoType() == AmmoType.T_NA))) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/battlevalue/ProtoMekBVCalculator.java:318]: Same here

**Review by Windchild292 — COMMENTED** [review]
This is a great improvement on the current setup, but carries some of the previous issues forwards. I've thus suggested a bunch of improvements on the moved but unchanged code.

File: megamek/src/megamek/common/battlevalue/AeroBVCalculator.java [hunk 6]
```diff
@@ -0,0 +1,1284 @@
+/*
+ * Copyright (c) 2022 - The MegaMek Team. All Rights Reserved.
+ *
+ * protoMek file is part of MegaMek.
```
[the remark this answers is not in the document]
Comment (SJuliez) [reply megamek/src/megamek/common/battlevalue/AeroBVCalculator.java:4]: LOL at one point I tried to do a replace this -> protoMek in the ProtoMekBVCalculator (which I moved first) and then copied over the copyright statement from there.

File: megamek/src/megamek/common/battlevalue/AeroBVCalculator.java [hunk 7]
```diff
@@ -0,0 +1,1284 @@
...
+
+        double targetingSystemBVMod = 1.0;
+
+        if (aero instanceof FixedWingSupport) {
+            if (aero.hasWorkingMisc(MiscType.F_ADVANCED_FIRECONTROL)) {
+                targetingSystemBVMod = 1.0;
+            } else if (aero.hasWorkingMisc(MiscType.F_BASIC_FIRECONTROL)) {
+                targetingSystemBVMod = .9;
+            } else {
+                targetingSystemBVMod = .8;
```
[the remark this answers is not in the document]
Comment (SJuliez) [reply megamek/src/megamek/common/battlevalue/AeroBVCalculator.java:485]: bug bear???
In German, such numbers are always written with the leading zero, so I dont mind correcting this...

File: megamek/src/megamek/common/battlevalue/GunEmplacementBVCalculator.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            EquipmentType etype = mounted.getType();
+
+            // don't count destroyed equipment
+            if (mounted.isDestroyed()) {
+                continue;
+            }
+
+            if (((etype instanceof WeaponType) && etype.hasFlag(WeaponType.F_AMS))
+                    || ((etype instanceof AmmoType) && (((AmmoType) etype)
+                    .getAmmoType() == AmmoType.T_AMS))
```
[the remark this answers is not in the document]
Comment (SJuliez) [reply megamek/src/megamek/common/battlevalue/GunEmplacementBVCalculator.java:48]: Absolutely. As I wrote, I simply did not touch anything as it wouldnt show up clearly on github. Although thinking about it now, it obviously would show up when in different commits.

File: megamek/src/megamek/common/battlevalue/GunEmplacementBVCalculator.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.common.battlevalue;
+
+import megamek.common.*;
+
+public class GunEmplacementBVCalculator extends BVCalculator {
+
+    public static int calculateBV(GunEmplacement gunEmplacement, boolean ignoreC3, boolean ignoreSkill, StringBuffer bvText) {
```
[the remark this answers is not in the document]
Comment (SJuliez) [reply megamek/src/megamek/common/battlevalue/GunEmplacementBVCalculator.java:25]: I dont think I'd want to do that.

File: megamek/src/megamek/common/battlevalue/JumpShipBVCalculator.java [hunk 10]
```diff
@@ -0,0 +1,888 @@
...
+import megamek.common.weapons.bayweapons.BayWeapon;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class JumpShipBVCalculator extends BVCalculator {
+
+    public static int calculateBV(Jumpship jumpShip, boolean ignoreC3, boolean ignoreSkill, StringBuffer bvText) {
```
[the remark this answers is not in the document]
Comment (SJuliez) [reply megamek/src/megamek/common/battlevalue/JumpShipBVCalculator.java:31]: Ditto.

**Comment by SJuliez** [comment]
OK everything in (almost, please lets leave a line between class and first method). I could now continue work in this PR but there's not much of a point to that so if it is approved I'd just merge it and work in a new branch.

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**